shared/src/main/scala/kiama/util/Source.scala

```diff
@@ -106,8 +108,29 @@ case class BufferSource(contents: GapBuffer, name: String = "") extends Source {
  * A source that is a named file.
  */
 case class FileSource(name: String, encoding: String = "UTF-8") extends Source {
-
   val shortName = Filenames.dropCurrentPath(name)
-
   lazy val content = scala.io.Source.fromFile(name, encoding).mkString
+
+  // Custom equals and hashCode to compare FileSource instances
+  //
+  // This is necessary to ensure that the same FileSource is recognized even if it is referenced with different
+  // filenames (such as "foo.effekt", "./foo.effekt", or "/abs/path/to/foo.effekt")
+
+  // The canonical (absolute, symlink-resolved) file
+  private lazy val canonicalFile: File =
+    new File(name).getCanonicalFile
+
+  override def canEqual(other: Any): Boolean = other.isInstanceOf[FileSource]
+
+  override def equals(other: Any): Boolean = other match {
+    case that: FileSource =>
+      this.canonicalFile == that.canonicalFile &&
+      this.encoding      == that.encoding
+    case _ =>
+      false
+  }
+
+  override def hashCode(): Int =
+    // combine canonical path and encoding
+    (canonicalFile, encoding).##
```

Comment:
Funny, I have never seen this operator.

Reply:
ChatGPT proposed it ;)

Reply:
Is this just `.hashCode`?
```suggestion
    (canonicalFile, encoding).hashCode
```